src/Polly.Extensions/Telemetry/ResilienceTelemetryDiagnosticSource.cs

```diff
@@ -0,0 +1,123 @@
+    {
+        LoggerFactory = options.LoggerFactory;
+        Enrichers = options.Enrichers.ToList();
+        OutcomeFormatter = options.OutcomeFormatter;
+
+        _logger = options.LoggerFactory.CreateLogger(TelemetryUtil.PollyDiagnosticSource);
+        _counter = Meter.CreateCounter<int>(
+            "resilience-events",
+            description: "Tracks the number of important resilience events that occurred in resilience strategies.");
+    }
+
+    public ILoggerFactory LoggerFactory { get; }
+
+    public Func<Outcome, string> OutcomeFormatter { get; }
+
+    public List<Action<EnrichmentContext>> Enrichers { get; }
+
+    public override bool IsEnabled(string name) => true;
+
+    public override void Write(string name, object? value)
+    {
+        if (value is not TelemetryEventArguments args)
+        {
+            return;
+        }
+
+        LogEvent(args);
+        MeterEvent(args);
+    }
+
+    private void MeterEvent(TelemetryEventArguments args)
+    {
+        var context = EnrichmentContext.Get(args.Arguments, args.Outcome);
+
+        var tags = new TagList
+        {
+            { "event-name", args.EventName },
+            { "builder-name", args.Source.BuilderName },
+            { "strategy-name", args.Source.StrategyName },
+            { "strategy-type", args.Source.StrategyType },
+            { "result-type", args.Context.IsVoid ? "void" : args.Context.ResultType.Name.ToString(CultureInfo.InvariantCulture) }
+        };
+
+        if (args.Source.BuilderProperties.TryGetValue(TelemetryUtil.StrategyKey, out var key))
+        {
+            tags.Add("strategy-key", key);
+        }
+        else
+        {
+            tags.Add("strategy-key", null);
+        }
+
+        tags.Add("exception-name", args.Outcome?.Exception?.GetType().FullName);
+
+        foreach (var enricher in Enrichers)
+        {
+            enricher(context);
+        }
+
+        foreach (var pair in context.Tags)
+        {
+            tags.Add(pair.Key, pair.Value);
+        }
+
+        _counter.Add(1, tags);
+
+        EnrichmentContext.Return(context);
+    }
+
+    private void LogEvent(TelemetryEventArguments args)
+    {
+        string outcomeString = "null";
+
+        if (args.Outcome is Outcome outcome)
+        {
+            outcomeString = OutcomeFormatter(outcome);
+        }
+
+        if (!args.Source.BuilderProperties.TryGetValue(TelemetryUtil.StrategyKey, out var strategyKey))
+        {
+            strategyKey = "null";
+        }
+
+        if (args.Outcome?.Exception != null)
+        {
+            ResilienceEventLog(_logger, args.EventName, args.Source.BuilderName, args.Source.StrategyName, args.Source.StrategyType, strategyKey, outcomeString, args.Outcome.Value.Exception);
+        }
+        else
+        {
+            ResilienceEventLog(_logger, args.EventName, args.Source.BuilderName, args.Source.StrategyName, args.Source.StrategyType, strategyKey, outcomeString, null);
+        }
```

Comment:
What's the difference between `args.Outcome.Exception` and `args.Outcome.Value.Exception`?

Reply:
Changed to make this clearer.